File: kluctl/utils/pretty_table.py
from kluctl.utils.utils import runHelper


def pretty_table(table, limit_widths):
    cols = len(table[0])

    def max_width(col, max_w):
        w = 0
        for l in table:
            w = max(w, len(l[col]))
        if max_w != -1:
            w = min(w, max_w)
        return w

    widths = [max_width(i, limit_widths[i] if i < len(limit_widths) else -1) for i in range(cols)]

    if len(limit_widths) < cols:
        # last column should use all remaining space
        try:
            # Try to figure out terminal width
            _, stdout, _ = runHelper(['stty', 'size'], stderr_log_level=None)
            _, term_columns = stdout.split()
            term_columns = int(term_columns)
        except:
            # Probably not a terminal, so let's use some large default
            term_columns = 200
        widths[len(limit_widths)] = term_columns - sum(widths[:-1]) - (cols - 1) * 3 - 4

    horizontal_separator = '+-'
    for i in range(cols):
        horizontal_separator += '-' * widths[i]
        if i != cols - 1:
            horizontal_separator += '-+-'
    horizontal_separator += '-+'

    ret = ''
    ret += horizontal_separator + '\n'
    for l in table:
        pos = [0] * cols

        while any([pos[i] < len(l[i]) for i in range(cols)]):
            ret += '| '
            for i in range(cols):
                t = l[i][pos[i]:pos[i] + widths[i]]
                new_line = t.find('\n')
                if new_line != -1:
                    t = t[:new_line]
                    pos[i] += 1
                pos[i] += len(t)
                ret += t
                ret += ' ' * (widths[i] - len(t))
                if i != cols - 1:
                    ret += ' | '
            ret += ' |\n'

        ret += horizontal_separator + '\n'

    return ret

File: pkg/utils/prettytable.go
package utils

import (
	"bytes"
	"github.com/kluctl/kluctl/lib/term"
	"sort"
	"strings"
	"unicode/utf8"
)

type Row []string

type PrettyTable struct {
	rows []Row
}

func (t *PrettyTable) AddRow(c ...string) {
	// expand tabs to 4 spaces, otherwise formatting breaks
	formatted := make([]string, len(c))
	for i, str := range c {
		formatted[i] = strings.ReplaceAll(str, "\t", strings.Repeat(" ", 4))
	}
	t.rows = append(t.rows, formatted)
}

func (t *PrettyTable) SortRows(col int) {
	sort.SliceStable(t.rows[1:], func(i, j int) bool {
		return t.rows[i+1][col] < t.rows[j+1][col]
	})
}

func (t *PrettyTable) Render(limitWidths []int) string {
	cols := len(t.rows[0])

	maxWidth := func(col int, maxW int) int {
		w := 0
		for _, l := range t.rows {
			for _, cl := range strings.Split(l[col], "\n") {
				if utf8.RuneCountInString(cl) > w {
					w = utf8.RuneCountInString(cl)
				}
			}
		}
		if maxW != -1 {
			if maxW < w {
				w = maxW
			}
		}
		return w
	}
	subStr := func(str string, s int, e int) string {
		if s > utf8.RuneCountInString(str) {
			s = utf8.RuneCountInString(str)
		}
		if e > utf8.RuneCountInString(str) {
			e = utf8.RuneCountInString(str)
		}
		return string(([]rune(str))[s:e])
	}

	widths := make([]int, cols)
	widthSum := 0
	for i := 0; i < cols; i++ {
		w := -1
		if i < len(limitWidths) {
			w = limitWidths[i]
		}
		widths[i] = maxWidth(i, w)
		if i != cols-1 {
			widthSum += widths[i]
		}
	}

	if len(limitWidths) < cols {
		tw := term.GetWidth()
		// last column should use all remaining space
		tw = tw - widthSum - (cols-1)*3 - 4
		if tw <= 0 {
			tw = 1
		}
		widths[len(limitWidths)] = tw
	}

	hsep := "+-"
	for i := 0; i < cols; i++ {
		hsep += strings.Repeat("-", widths[i])
		if i != cols-1 {
			hsep += "-+-"
		}
	}
	hsep += "-+\n"

	buf := bytes.NewBuffer(nil)
	buf.WriteString(hsep)
	pos := make([]int, cols)
	for _, l := range t.rows {
		for i := 0; i < cols; i++ {
			pos[i] = 0
		}

		for {
			anyLess := false
			for i := 0; i < cols; i++ {
				if pos[i] < utf8.RuneCountInString(l[i]) {
					anyLess = true
				}
			}
			if !anyLess {
				break
			}

			buf.WriteString("| ")
			for i := 0; i < cols; i++ {
				x := subStr(l[i], pos[i], pos[i]+widths[i])
				newLine := strings.IndexRune(x, '\n')
				if newLine != -1 {
					x = x[:newLine]
					pos[i] += 1
				}
				pos[i] += utf8.RuneCountInString(x)
				buf.WriteString(x)
				buf.WriteString(strings.Repeat(" ", widths[i]-utf8.RuneCountInString(x)))
				if i != cols-1 {
					buf.WriteString(" | ")
				}
			}
			buf.WriteString(" |\n")
		}
		buf.WriteString(hsep)
	}
	return buf.String()
}
